servers/src/main/java/tachyon/master/MasterBase.java

```
@@ -0,0 +1,166 @@
+
+import com.google.common.base.Preconditions;
+
+import tachyon.Constants;
+import tachyon.master.journal.Journal;
+import tachyon.master.journal.JournalEntry;
+import tachyon.master.journal.JournalInputStream;
+import tachyon.master.journal.JournalSerializable;
+import tachyon.master.journal.JournalTailer;
+import tachyon.master.journal.JournalTailerThread;
+import tachyon.master.journal.JournalWriter;
+
+public abstract class MasterBase implements Master {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private final Journal mJournal;
+  private final ExecutorService mExecutorService;
+
+  /** true if this master is in leader mode, and not standby mode. */
+  private boolean mIsLeader = false;
+
+  /** The thread that tails the journal when the master is in standby mode. */
+  private JournalTailerThread mStandbyJournalTailer = null;
+
+  private JournalWriter mJournalWriter = null;
+
+  protected MasterBase(Journal journal, ExecutorService executorService) {
+    mJournal = Preconditions.checkNotNull(journal);
+    mExecutorService = Preconditions.checkNotNull(executorService);
+  }
+
+  @Override
+  public void processJournalCheckpoint(JournalInputStream inputStream) throws IOException {
+    JournalEntry entry;
+    while ((entry = inputStream.getNextEntry()) != null) {
+      processJournalEntry(entry);
+    }
+    inputStream.close();
+  }
+
+  @Override
+  public void start(boolean isLeader) throws IOException {
+    LOG.info("Starting master. isLeader: " + isLeader);
+    mIsLeader = isLeader;
+    if (mIsLeader) {
+      // Replay all the state of the checkpoint and the completed log files.
+      // TODO: only do this if this is a fresh start, not if this master had already been tailing
+      // the journal.
+      LOG.info(getProcessorName() + ": process completed logs before becoming master.");
+      JournalTailer catchupTailer = new JournalTailer(this, mJournal);
+      boolean checkpointExists = true;
```

It may be better to avoid this boolean and just put the logic in the `if` clause into the `try` and just ignore the exception in the `catch`.